apis/kubedb/v1alpha2/kafka_helpers.go

```diff
@@ -0,0 +1,318 @@
+	return k.OffshootName()
+}
+
+func (k *Kafka) GoverningServiceName() string {
+	return meta_util.NameWithSuffix(k.ServiceName(), "pods")
+}
+
+func (k *Kafka) GoverningServiceNameController() string {
+	return meta_util.NameWithSuffix(k.ServiceName(), KafkaNodeRolesController)
+}
+
+func (k *Kafka) GoverningServiceNameBroker() string {
+	return meta_util.NameWithSuffix(k.ServiceName(), KafkaNodeRolesBrokers)
+}
+
+func (k *Kafka) StandbyServiceName() string {
+	return meta_util.NameWithPrefix(k.ServiceName(), "standby")
+}
+
+func (k *Kafka) offshootLabels(selector, override map[string]string) map[string]string {
+	selector[meta_util.ComponentLabelKey] = ComponentDatabase
+	return meta_util.FilterKeys(kubedb.GroupName, selector, meta_util.OverwriteKeys(nil, k.Labels, override))
+}
+
+func (k *Kafka) OffshootSelectors(extraSelectors ...map[string]string) map[string]string {
+	selector := map[string]string{
+		meta_util.NameLabelKey:      k.ResourceFQN(),
+		meta_util.InstanceLabelKey:  k.Name,
+		meta_util.ManagedByLabelKey: kubedb.GroupName,
+	}
+	return meta_util.OverwriteKeys(selector, extraSelectors...)
+}
+
+func (k *Kafka) ControllerNodeSelectors() map[string]string {
+	return meta_util.OverwriteKeys(k.OffshootSelectors(), map[string]string{
+		k.NodeRoleSpecificLabelKey(KafkaNodeRoleController): KafkaNodeRoleSet,
+	})
+}
+
+func (k *Kafka) BrokerNodeSelectors() map[string]string {
+	return meta_util.OverwriteKeys(k.OffshootSelectors(), map[string]string{
+		k.NodeRoleSpecificLabelKey(KafkaNodeRoleBroker): KafkaNodeRoleSet,
+	})
+}
+
+func (k *Kafka) OffshootLabels() map[string]string {
+	return k.offshootLabels(k.OffshootSelectors(), nil)
+}
+
+func (k *Kafka) ServiceLabels(alias ServiceAlias, extraLabels ...map[string]string) map[string]string {
+	svcTemplate := GetServiceTemplate(k.Spec.ServiceTemplates, alias)
+	return k.offshootLabels(meta_util.OverwriteKeys(k.OffshootSelectors(), extraLabels...), svcTemplate.Labels)
+}
+
+func (k *Kafka) ControllerServiceLabels() map[string]string {
+	return meta_util.OverwriteKeys(k.offshootLabels(k.OffshootLabels(), k.ControllerNodeSelectors()))
+}
+
+func (k *Kafka) BrokerServiceLabels() map[string]string {
+	return meta_util.OverwriteKeys(k.offshootLabels(k.OffshootLabels(), k.BrokerNodeSelectors()))
+}
+
+func (k *Kafka) PodControllerLabels(extraLabels ...map[string]string) map[string]string {
+	return k.offshootLabels(meta_util.OverwriteKeys(k.OffshootSelectors(), extraLabels...), k.Spec.PodTemplate.Controller.Labels)
+}
+
+func (k *Kafka) PodLabels(extraLabels ...map[string]string) map[string]string {
+	return k.offshootLabels(meta_util.OverwriteKeys(k.OffshootSelectors(), extraLabels...), k.Spec.PodTemplate.Labels)
+}
+
+func (k *Kafka) StatefulSetName() string {
+	return k.OffshootName()
+}
+
+func (k *Kafka) CombinedStatefulSetName() string {
+	return k.StatefulSetName()
+}
+
+func (k *Kafka) ControllerStatefulSetName() string {
+	if k.Spec.Topology.Controller.Suffix != "" {
```

Comment:
check `k.Spec.Topology != nil` first

Reply:
Not required. This method is only called when `k.Spec.Topology != nil`